packages/lodestar/src/chain/validation/attestation.ts

```diff
@@ -1,177 +1,286 @@
+  const latestPermissibleSlot = chain.clock.slotWithTolerance(MAXIMUM_GOSSIP_CLOCK_DISPARITY_SEC);
+  const earliestPermissibleSlot = Math.max(
+    // slot with past tolerance of MAXIMUM_GOSSIP_CLOCK_DISPARITY_SEC
+    // ATTESTATION_PROPAGATION_SLOT_RANGE = SLOTS_PER_EPOCH
+    chain.clock.slotWithTolerance(-MAXIMUM_GOSSIP_CLOCK_DISPARITY_SEC) - SLOTS_PER_EPOCH,
+    0
+  );
+  if (attestationSlot < earliestPermissibleSlot) {
+    throw new AttestationError({code: AttestationErrorCode.PAST_SLOT, earliestPermissibleSlot, attestationSlot});
+  }
+  if (attestationSlot > latestPermissibleSlot) {
+    throw new AttestationError({code: AttestationErrorCode.FUTURE_SLOT, latestPermissibleSlot, attestationSlot});
   }
+}
+
+/**
+ * Verify:
+ * 1. head block is known
+ * 2. attestation's target block is an ancestor of the block named in the LMD vote
+ */
+export function verifyHeadBlockAndTargetRoot(
+  chain: IBeaconChain,
+  beaconBlockRoot: Root,
+  targetRoot: Root,
+  attestationEpoch: Epoch
+): void {
+  const headBlock = verifyHeadBlockIsKnown(chain, beaconBlockRoot);
+  verifyAttestationTargetRoot(headBlock, targetRoot, attestationEpoch);
+}
+
+/**
+ * Checks if the `attestation.data.beaconBlockRoot` is known to this chain.
+ *
+ * The block root may not be known for two reasons:
+ *
+ * 1. The block has never been verified by our application.
+ * 2. The block is prior to the latest finalized block.
+ *
+ * Case (1) is the exact thing we're trying to detect. However case (2) is a little different, but
+ * it's still fine to reject here because there's no need for us to handle attestations that are
+ * already finalized.
+ */
+function verifyHeadBlockIsKnown(chain: IBeaconChain, beaconBlockRoot: Root): IBlockSummary {
+  // TODO (LH): Enforce a maximum skip distance for unaggregated attestations.
 
-  if (!chain.forkChoice.isDescendantOfFinalized(attestation.data.beaconBlockRoot)) {
+  const headBlock = chain.forkChoice.getBlock(beaconBlockRoot);
+  if (headBlock === null) {
     throw new AttestationError({
-      code: AttestationErrorCode.FINALIZED_CHECKPOINT_NOT_AN_ANCESTOR_OF_ROOT,
-      job: attestationJob,
+      code: AttestationErrorCode.UNKNOWN_BEACON_BLOCK_ROOT,
+      root: beaconBlockRoot as Uint8Array,
     });
   }
-  db.seenAttestationCache.addCommitteeAttestation(attestation);
 
-  return indexedAttestation;
+  return headBlock;
 }
 
-export function getAttestationAttesterCount(attestation: phase0.Attestation): number {
-  return Array.from(attestation.aggregationBits).filter((bit) => !!bit).length;
-}
+/**
+ * Verifies that the `attestation.data.target.root` is indeed the target root of the block at
+ * `attestation.data.beacon_block_root`.
+ */
+function verifyAttestationTargetRoot(headBlock: IBlockSummary, targetRoot: Root, attestationEpoch: Epoch): void {
+  // Check the attestation target root.
+  const headBlockEpoch = computeEpochAtSlot(headBlock.slot);
+
+  if (headBlockEpoch > attestationEpoch) {
```

Comment:
this does not go with the comment, please double check

Reply:
ok so headBlock here is not the head of our forkchoice, the name is misleading

Reply:
What name would you use here?

Reply:
just read it again, `headBlock` means the head block the attester voted for so it's still valid, but in lodestar we mean the current head of our forkchoice
maybe just adding a comment in the method header would be enough